### PR TITLE
implement time

### PR DESCRIPTION
**Pull request by kaplanelad**
implement time

### CODE REVIEW

**Comment by kaplanelad** [comment]
Hey @sandhose i have started with time implementation.
I just finished with all the function except `time.parse_ns` and `time.diff`

I would like to hear your opinion before implementation
1. `time.now_ns` - is a dynamic response, how did you plan to add the value in snapshot? (now it mark off from the test rego
2. `time.parse_ns` - to implement this function I need to use the same implementation like go has with the same format time? you expect to implement all this function under time or i can separate to different pr?

File: src/builtins/impls/time.rs [hunk 1]
```diff
@@ -14,67 +14,147 @@
...
+        Some(_data) => match (value.get(0).unwrap_or(value).as_i64(), value.get(1)) {
+            (Some(ns), Some(timezone)) => (ns, timezone.as_str().unwrap_or_default().to_string()),
+            _ => return Err(anyhow!(ERROR_INVALID_ARGUMENTS)),
+        },
+        _ => match value.as_i64() {
+            Some(ns) => (ns, DEFAULT_TIMEZONE.to_string()),
+            None => return Err(anyhow!(ERROR_INVALID_ARGUMENTS)),
+        },
+    };
+    nanoseconds_to_date(ns, timezone.as_str())
```
Comment (sandhose) [inline src/builtins/impls/time.rs:159]: If I used `serde_json::Value` as builtin arguments is that I was lazy and did not defined a custom type.

What you can do instead is define a type like this:

```rust
#[derive(Serialize, Deserialize)]
#[serde(untagged)]
enum TimestampWithOptionalTimezone {
    Timestamp(i64),
    TimestampAndTimezone(i64, String),
}
```

File: src/builtins/impls/time.rs [hunk 2]
```diff
@@ -14,67 +14,147 @@
...
+        Weekday::Mon => "Monday",
+        Weekday::Tue => "Tuesday",
+        Weekday::Wed => "Wednesday",
+        Weekday::Thu => "Thursday",
+        Weekday::Fri => "Friday",
+        Weekday::Sat => "Saturday",
+        Weekday::Sun => "Sunday",
+    }
+    .to_string())
+}
```
Comment (sandhose) [inline src/builtins/impls/time.rs:108]: You can also also return a `&'static str` instead here and avoid the `to_string()`

File: src/builtins/impls/time.rs [hunk 3]
```diff
@@ -14,67 +14,147 @@
...
+
+/// convert nanoseconds to [`chrono::DateTime`] by the given timezone
+///
+/// # Errors
+///
+/// - the `time_zone` when could not parse the given timezone
+/// - the given `ns` representation is invalid.
+fn nanoseconds_to_date(ns: i64, time_zone: &str) -> Result<DateTime<Tz>> {
```
Comment (sandhose) [inline src/builtins/impls/time.rs:116]: Here I'd make `time_zone` an `Option<&str>` and do something like

```rust
let tz = time_zone
  .map(|tz| tz.parse())
  .transpose()?
  .unwrap_or(Tz::UTC);
```


which avoids having to parse `"UTC"` or having to define a `DEFAULT_TIMEZONE` const

**Review by sandhose — CHANGES_REQUESTED** [review]
That's impressive, thanks a lot!

I think the main change I'd like to see is to define a custom type for functions that take either a `ns` or a `[ns, tz]` argument instead of using a `serde_json::Value`

File: src/builtins/impls/time.rs (hunk 1 above)
Comment (kaplanelad) [reply src/builtins/impls/time.rs:159]: you mean that the function `nanoseconds_to_date` will receive `TimestampWithOptionalTimezone` enum
```rs
fn nanoseconds_to_date(timestamp: TimestampWithOptionalTimezone) -> Result<DateTime<Tz>> {
...
}
```

And replace `let (ns, timezone) = match value.as_array() {...` to  `let timestamp: TimestampWithOptionalTimezone = match value.as_array() {...`

Comment (sandhose) [reply src/builtins/impls/time.rs:159]: Not exactly: builtins are implemented in a way where parameters can be anything that can be deserialised by serde.

This means you could define this `TimestampWithOptionalTimezone` and use it directly in the builtins parameters like

```
#[derive(Serialize, Deserialize)]
#[serde(untagged)]
enum TimestampWithOptionalTimezone {
    Timestamp(i64),
    TimestampAndTimezone(i64, String),
}

impl TimestampWithOptionalTimezone {
    fn to_datetime(self) -> Result<Datetime<Tz>> {
        match self {
            Timestamp(ts) => todo!(),
            TimestampAndTimezone(ts, tz) => todo!(),
        }
    }
} 

pub fn weekday(x: TimestampWithOptionalTimezone) -> Result<&'static str> {
    let date_time = x.to_datetime()?;
    Ok(match date_time.weekday() {
        Weekday::Mon => "Monday",
        Weekday::Tue => "Tuesday",
        Weekday::Wed => "Wednesday",
        Weekday::Thu => "Thursday",
        Weekday::Fri => "Friday",
        Weekday::Sat => "Saturday",
        Weekday::Sun => "Sunday",
    })
}

**Comment by kaplanelad** [comment]
@sandhose, did your changes requests

**Comment by kaplanelad** [comment]
@sandhose any option that you can review this PR? I did all your changes.

**Comment by sandhose** [comment]
> time now function did't tested (commented out), this is a dynamic value what are the thoughts?

That should be passed in the evaluation context, same as the random number generator. I'll do that in another PR.

> `diff` and `parse_ns` not implemented yes. can we keep it to new new pr?

Sure!

---

I did rebase your branch, and did what I meant with the `TimestampWithOptionalTimezone`: use it as parameter, so that you don't have to deal with `serde_json::Value` manually

**Comment by kaplanelad** [comment]
Thanks @sandhose I'll check your commit and learn for the next time

**Review by sandhose — APPROVED** [review]

**Pull request merged**